### PR TITLE
cpu/lpc23xx: implement periph/flashpage

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

This converts the old flashrom driver to the new flashpage API.
I've had this lying around for a while and now all dependencies are merged, but unfortunately I can't test it.
Not sure if this works yet, but better PR it, maybe it's useful for someone. 

### Testing procedure

`tests/periph_flashpage` should now work on lpc23xx boards (untested)

### Issues/PRs references

follow-up to #16280
#15936 could come in handy here.

### CODE REVIEW

File: cpu/lpc23xx/periph/flashpage.c [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+ *
+ * Parameters:  long sect_start:    Param0
+ *              long sect_end:      Param1
+ *
+ * Return:      Code    CMD_SUCCESS
+ *                      BUSY
+ *                      SECTOR_NOT_BLANK
+ *                      INVALID_SECTOR
+ *
+ *              Result0: Offset of the first non blank word location if the status code is SECTOR_NOT_BLANK.
```
Comment (maribu) [inline cpu/lpc23xx/periph/flashpage.c:73]:
```suggestion
 *              Result0: Offset of the first non blank word location if the status
                         code is SECTOR_NOT_BLANK.
```

File: cpu/lpc23xx/periph/flashpage.c [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+ * Return:      Code    CMD_SUCCESS
+ *                      or SRC_ADDR_ERROR (Address not on word boundary)
+ *                      or DST_ADDR_ERROR (Address not on correct boundary)
+ *                      or SRC_ADDR_NOT_MAPPED
+ *                      or DST_ADDR_NOT_MAPPED
+ *                      or COUNT_ERROR (Byte count is not 256 | 512 | 1024 | 4096 | 8192)
+ *                      or SECTOR_NOT_PREPARED_FOR_WRITE_OPERATION
+ *                      or BUSY
+ *****************************************************************************/
+static uint32_t copy_ram_to_flash(void* dst, const void *src, size_t tmp_size)
```
Comment (maribu) [inline cpu/lpc23xx/periph/flashpage.c:108]:
```suggestion
static uint32_t copy_ram_to_flash(void *dst, const void *src, size_t tmp_size)
```

File: cpu/lpc23xx/periph/flashpage.c [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+    return iap(COPY_RAM_TO_FLASH, (uintptr_t)dst, (uintptr_t)src, tmp_size, _XTAL);
+}
+
+/******************************************************************************
+ * Function:    Prepare_Sector
+ *
+ * Description: This command must be executed before executing "Copy RAM to Flash" or "Erase Sector(s)"
+ *              command. Successful execution of the "Copy RAM to Flash" or "Erase Sector(s)" command causes
+ *              relevant sectors to be protected again. The boot sector can not be prepared by this command. To
+ *              prepare a single sector use the same "Start" and "End" sector numbers..
```
Comment (maribu) [inline cpu/lpc23xx/periph/flashpage.c:119]: Please break lines to fit into 100 chars

File: cpu/lpc23xx/periph/flashpage.c [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+{
+    return iap(PREPARE_SECTOR_FOR_WRITE_OPERATION, sect_start, sect_end, 0, 0);
+}
+
+/******************************************************************************
+ * Function:    erase_sectors
+ *
+ * Description: This command is used to erase a sector or multiple sectors of on-chip Flash memory. The boot
+ *              sector can not be erased by this command. To erase a single sector use the same "Start" and "End"
+ *              sector numbers.
```
Comment (maribu) [inline cpu/lpc23xx/periph/flashpage.c:141]: again 100 chars limit exceeded

File: cpu/lpc23xx/periph/flashpage.c [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+{
+    return iap(ERASE_SECTOR, sect_start, sect_end, _XTAL, 0);
+}
+
+/******************************************************************************
+ * Function:    compare
+ *
+ * Description: This command is used to compare the memory contents at two locations. compare result may not
+ *              be correct when source or destination address contains any of the first 64 bytes starting
+ *              from address zero. First 64 bytes can be re-mapped to RAM.
```
Comment (maribu) [inline cpu/lpc23xx/periph/flashpage.c:165]: Same

**Review by maribu — COMMENTED** [review]
looks good code-wise, only some minor style nits

**Comment by maribu** [comment]
@benpicco: Is this tested, or should I wire up my MSB-A2?

**Comment by benpicco** [comment]
It's untested, I no longer have access to the `mcb2388`

**Comment by maribu** [comment]
```
$ make BOARD=msba2 flash test -C tests/periph_flashpage

[...]

main(): This is RIOT! (Version: 2021.07-devel-93-g51d2d-ben)
ROM flash read write test

Please refer to the README.md for further information

Flash start addr:	0x00000000
Page size:		variable
Number of pages:	27
> 

> 
> test_last_raw

> test_last_raw
error verifying the content of last page
> Timeout in expect script at "child.expect_exact('wrote raw short buffer to last flash page')" (tests/periph_flashpage/tests/01-run.py:20)

make: *** [/home/maribu/Repos/software/RIOT/makefiles/tests/tests.inc.mk:22: test] Error 
```

:-(

**Comment by benpicco** [comment]
Yea I was expecting this.
The test implicitly expects `FLASHPAGE_WRITE_BLOCK_SIZE` to be no larger than 16 bytes, which has been true so far.
The flashpage API mandates to only write multiples of `FLASHPAGE_WRITE_BLOCK_SIZE` bytes at a time.

Now the flashpage driver could take care of the read-modify-write cycle to allow arbitrary writes (as it was done for sam0), but this then motivated me to do #15936.

**Comment by maribu** [comment]
Works now:

```
main(): This is RIOT! (Version: 2021.07-devel-94-gc7c4d-ben)
ROM flash read write test

Please refer to the README.md for further information

Flash start addr:	0x00000000
Page size:		variable
Number of pages:	27
> 

> 
> test_last_raw

> test_last_raw
wrote raw short buffer to last flash page
> help
help
Command              Description
---------------------------------------
info                 Show information about pages
write_raw            Write (ASCII, max 64B) data to the given address
erase                Erase the given page buffer
test_last_raw        Write and verify raw short write on last page available
> help
help
Command              Description
---------------------------------------
info                 Show information about pages
write_raw            Write (ASCII, max 64B) data to the given address
erase                Erase the given page buffer
test_last_raw        Write and verify raw short write on last page available
> 
```

**Review by maribu — APPROVED** [review]

**Pull request merged**